src/ucs/datastruct/piecewise_func.c

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2024. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <ucs/datastruct/piecewise_func.h>
+#include <ucs/debug/memtrack_int.h>
+#include <ucs/debug/assert.h>
+#include <limits.h>
+#include <stdint.h>
+
+ucs_status_t ucs_piecewise_func_init(ucs_piecewise_func_t *func)
+{
+    ucs_piecewise_segment_t *seg;
+    
+    seg = ucs_malloc(sizeof(ucs_piecewise_segment_t), "piecewise segment");
+    if (seg == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    seg->func = UCS_LINEAR_FUNC_ZERO;
+    seg->end  = SIZE_MAX;
+
+    ucs_list_head_init(&func->head);
+    ucs_list_insert_after(&func->head, &seg->list);
+
+    return UCS_OK;
+}
+
+void ucs_piecewise_func_cleanup(ucs_piecewise_func_t *func)
+{
+    ucs_piecewise_segment_t *seg, *tmp;
+
+    ucs_list_for_each_safe(seg, tmp, &func->head, list) {
+        ucs_free(seg);
+    }
+}
+
+static void
+ucs_piecewise_func_assert_covered_range(const ucs_piecewise_func_t *func)
+{
+#if ENABLE_ASSERT
+    ucs_piecewise_segment_t *tail = ucs_list_tail(&func->head,
+                                                  ucs_piecewise_segment_t,
+                                                  list);
+    ucs_piecewise_segment_t *seg;
+
+    ucs_list_for_each(seg, &func->head, list) {
+        if (seg->end == SIZE_MAX) {
+            ucs_assertv(seg == tail,
+                        "Segment that covers SIZE_MAX isn't the last one");
+            return;
+        }
+    }
+
+    ucs_bug("List doesn't cover [0, SIZE_MAX] range");
+#endif
+}
+
+double ucs_piecewise_func_apply(const ucs_piecewise_func_t *func, size_t x)
+{
+    ucs_piecewise_segment_t *seg;
+
+    ucs_piecewise_func_assert_covered_range(func);
+
+    ucs_list_for_each(seg, &func->head, list) {
+        if (seg->end >= x) {
+            break;
+        }
+    }
```

Comment:
unite common code with other functions - create a helper function to find the segment that contains point x